torch/csrc/jit/codegen/cuda/lower_utils.cpp

```diff
@@ -277,6 +337,195 @@ std::pair<kir::ForLoop*, int64_t> getAllocPoint(
+    for (auto new_expr : replaced_then_body) {
+      new_ite->thenBody().push_back(new_expr);
+    }
+    if (ite->hasElse()) {
+      auto replaced_else_body =
+          replace(ite->elseBody().exprs(), replacement_map_);
+      for (auto new_expr : replaced_else_body) {
+        new_ite->elseBody().push_back(new_expr);
+      }
+    }
+    replaced_expr_ = new_ite;
+  }
+
+  void visit(kir::UnaryOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::UnaryOp>(
+          node->operation(),
+          node->out(),
+          replaced_inputs.value().at(node->in()));
+    }
+  }
+  void visit(kir::BinaryOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::BinaryOp>(
+          node->operation(),
+          node->out(),
+          replaced_inputs.value().at(node->lhs()),
+          replaced_inputs.value().at(node->rhs()));
+    }
+  }
+
+  void visit(kir::TernaryOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::TernaryOp>(
+          node->operation(),
+          node->out(),
+          replaced_inputs.value().at(node->in1()),
+          replaced_inputs.value().at(node->in2()),
+          replaced_inputs.value().at(node->in3()));
+    }
+  }
+
+  void visit(kir::ReductionOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::ReductionOp>(
+          node->operation(),
+          node->init(),
+          node->out(),
+          replaced_inputs.value().at(node->in()));
+    }
+  }
+
+  void visit(kir::BroadcastOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::BroadcastOp>(
+          node->out(), replaced_inputs.value().at(node->in()));
+    }
+  }
+
+  void visit(kir::WelfordOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::WelfordOp>(
+          node->outAvg(),
+          node->outVar(),
+          node->outN(),
+          node->initAvg(),
+          node->initVar(),
+          node->initN(),
+          replaced_inputs.value().at(node->inAvg()),
+          replaced_inputs.value().at(node->inVar()),
+          replaced_inputs.value().at(node->inN()));
+    }
+  }
+
+ private:
+  GpuLower* gpu_lower_;
+  kir::IrBuilder ir_builder_;
+  kir::Expr* replaced_expr_ = nullptr;
+  const std::unordered_map<kir::Val*, kir::Val*>& replacement_map_;
+};
+
+} // namespace
+
+std::vector<kir::Expr*> replaceInputsInExpr(
```

Comment:
There's also `ir_utils::applyReplacements`, although it's only for expressions. It seems we should have some generic find-and-replace interface for KIR.